### ISSUE TITLE
Type narrowing breaks type check

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#9955]
**Describe the bug**
I think I am having a conceptual issue here - at least mypy is giving the same error:

> bug.py:7:12 - error: Type "int" is not assignable to return type "T@factory"
    Type "int" is not assignable to type "T@factory" (reportReturnType)
1 error, 0 warnings, 0 informations

> bug.py:7: error: Incompatible return value type (got "int", expected "T")  [return-value]
Found 1 error in 1 file (checked 1 source file)

What I find remarkable is that when removing the `assert` line, the code type-checks as expected. But narrowing down `my_type` seems to break the connection between `my_type` and `T`, although obviously, `my_type` is compatible with `T` and should be seen as such.

**Code or Screenshots**
```python
from typing import Any


def factory[T: Any](my_type: type[T]) -> T:
    assert issubclass(my_type, int)

    return my_type(1)
```

**VS Code extension or command-line**
pyright 1.1.394 from PyPI

**Comment by erictraut** [comment]
You should never use `Any` as an upper bound. The default upper bound is `object`, which effectively imposes no constraint on the type variable. There's _never_ a good reason to specify `Any` or `object` as an upper bound, and the typing spec should probably be updated to indicate that it's an error condition to use `Any` as an upper bound. If you fix that in your code, the sample type checks fine in pyright.

With mypy, the problem is that it narrows the type of `my_type` to `type[int]` whereas pyright narrows its type to `type[int]*`. The asterisk denotes a [conditional type](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables).
